### PR TITLE
Support for Mixed Input TensorOp 

### PR DESCRIPTION
**Pull request by manishucsd**
This PR introduces support for mixed input data types for the NVIDIA A100. The PR is motivated by rising trend of utilizing activations in 16-bit floating-point (either F16 or BF16) alongside weights in 8-bit integer. 

# Supported GEMM datatype configurations
### Upcast on Operand B
- `[F16|F32] <= F16 * S8 + [F16|F32]`
- `[F16|F32] <= F16 * U8 + [F16|F32]`
### Upcast on Operand A
- `[F16|F32] <= S8 * F16 + [F16|F32]`
- `[F16|F32] <= U8 * F16 + [F16|F32]`

### Upcast on Operand B
- `[BF16|F32] <= BF16 * S8 + F32`
- `[BF16|F32] <= BF16 * U8 + F32`
### Upcast on Operand A
- `[BF16|F32] <= S8 * BF16 + F32`
- `[BF16|F32] <= U8 * BF16 + F32`

## Key Features of the Mixed Input Computation:

- **Memory Layout:** TN layout, which implies RowMajor for matrix A and ColumnMajor for matrix B when they're in the Global Memory.
  
- **Mainloop Execution Blueprint:** 
    1. **Data Loading:** Copies data from Global Memory to Shared Memory (using either `cp.async` [or `cp.async.bulk`]. `cp.async`(A100 version) is implemented in this PR).
    2. **Shared Memory to Register:** Loads data from Shared Memory to Registers through the `ldmatrix` instruction.
    3. **Layout Adaptation for `mma` Operation:** 
        - Incorporates the `FragmentShuffler` component, which efficiently reorganizes data within a warp to align with the thread-value mapping needed by the `mma` operation. 
        - Given that `ldmatrix` operates on 8-bit values (`S8` or `U8`) and `mma.sync` works on 16-bit values (`F16` or `BF16`), there's a mismatch in thread ownership post `ldmatrix`. The data isn't aligned with what `mma.sync.[f16|f32].f16.f16.[f16|f32]` or `mma.sync.f32.bf16.bf16.f32` expects. The `FragmentShuffler` addresses this challenge. As all necessary data is within the warp (32 threads), `FragmentShuffler` shuffles it to achieve a thread layout that conforms to `mma.sync`.
      4. **Data Type Conversion:** Adapts data to align with the wider of the two datatype, making it suitable for processing by the math instruction. This is facilitated by the `FragmentConverter`, leveraging the efficiencyt and newly added`FastNumericArrayConverter` for each of the four conversion combinations [`4xS8` | `4xU8`] -> [`4xF16`|`4xBF16`].

## Performance Profiling 
GEMM Performance with Mixed-Input Datatype on NVIDIA A100 40GB SXM4

The top commit on this PR archives the performance as shown by the graph above for the various mixed-input datatype on an NVIDIA A100 40GB SXM4 Chip

### Reproducing Performance Results
In addition to the implementation, the kernels are also added to the profiler. To repro the above performance results, follow the following steps:
```
# cmake to generate only the kernels that we want to profile
build $ cmake ../cutlass/ -DCUTLASS_NVCC_ARCHS='80' -DCUTLASS_LIBRARY_KERNELS="\
s16816gemm_f16_s8_128x128_64x3,\
s16816gemm_s8_f16_128x128_64x3,\
s16816gemm_u8_f16_128x128_64x3,\
s16816gemm_f16_u8_128x128_64x3,\
s16816gemm_bf16_s8_128x128_64x3,\
s16816gemm_s8_bf16_128x128_64x3,\
s16816gemm_bf16_u8_128x128_64x3,\
s16816gemm_u8_bf16_128x128_64x3,\
tensorop_s16816gemm_f16_128x128_64x*_tn_align8,\
tensorop_s16816gemm_bf16_128x128_64x*_tn_align8" -DCUTLASS_NVCC_KEEP=OFF -DCUTLASS_LIBRARY_IGNORE_KERNELS="gemm_grouped*,gemm_planar*,f16_s16816gemm_f16_128x128_64x4_tn_align8"

# build cutlass_profiler and stuff affected by this PR
build $ make cutlass_test_unit_core cutlass_test_unit_gemm_warp cutlass_test_unit_gemm_device_mixed_input_tensorop_sm80 cutlass_profiler -j

# On NVIDIA A100-SXM4-40GB - SM 8.0, 108 SMs @ 1410 MHz, L2 cache: 40 MB, Global Memory: 39 GB
build $ ./tools/profiler/cutlass_profiler --op_class=tensorop --m=3456 --n=4096 --k=8192 --output=cutlass_data.csv
```

# Breakdown of Performance Improvements
Mixed Input Data Type (F16 _= F16 _ U8 + F16) Performance Progressing with Increasing Mainloop Optimizations

The figure above shows the performance improvements for one mixed input datatype with increasing level of mainloop optimization. The idea is to avoid long-latency memory instructions, and long sequence of arithmetic conversion instructions inside the mainloop.

## Functional Testing
- Added device level tests.

- Added warp-level tests for each variant. Tests the `FragmentShuffler` and `FragmentConverter`. 
```bash
build $ make cutlass_test_unit_gemm_warp -j

$ ./test/unit/gemm/warp/cutlass_test_unit_gemm_warp --gtest_filter=SM80_warp_gemm_mixed_input_tensor_op*
Note: Google Test filter = SM80_warp_gemm_mixed_input_tensor_op*
[==========] Running 5 tests from 3 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_i8
[ RUN      ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_i8.128x128x64_64x64x64_16x8x16
[       OK ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_i8.128x128x64_64x64x64_16x8x16 (155 ms)
[ RUN      ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_i8.64x64x64_64x64x64_16x8x16
[       OK ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_i8.64x64x64_64x64x64_16x8x16 (3 ms)
[----------] 2 tests from SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_i8 (158 ms total)

[----------] 2 tests from SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_u8
[ RUN      ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_u8.64x64x64_64x64x64_16x8x16
[       OK ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_u8.64x64x64_64x64x64_16x8x16 (3 ms)
[ RUN      ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_u8.128x128x64_64x64x64_16x8x16
[       OK ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_u8.128x128x64_64x64x64_16x8x16 (4 ms)
[----------] 2 tests from SM80_warp_gemm_mixed_input_tensor_op_crosswise_f16_u8 (7 ms total)

[----------] 1 test from SM80_warp_gemm_mixed_input_tensor_op_crosswise_bf16_u8
[ RUN      ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_bf16_u8.64x64x64_64x64x64_16x8x16
[       OK ] SM80_warp_gemm_mixed_input_tensor_op_crosswise_bf16_u8.64x64x64_64x64x64_16x8x16 (2 ms)
[----------] 1 test from SM80_warp_gemm_mixed_input_tensor_op_crosswise_bf16_u8 (2 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 3 test suites ran. (169 ms total)
[  PASSED  ] 5 tests.

```
If this fails most-likely the issue will be in the warp-level components. Specifically in the `FragmentShuffler`. Note that any bugs in `FragmentConverter` will be caught in the newly added `test/core` described in the below bullet. 

- Added new test for `FastNumericConverter`. For 8b integer to 16b floating point, it tests all 8bit integer patterns which are not that many :).
```
build $ ./test/unit/core/cutlass_test_unit_core --gtest_filter=FastNumericConversion.*
Note: Google Test filter = FastNumericConversion.*
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from FastNumericConversion
[ RUN      ] FastNumericConversion.s32_to_f32
[       OK ] FastNumericConversion.s32_to_f32 (152 ms)
[ RUN      ] FastNumericConversion.s8_to_f16_array
[       OK ] FastNumericConversion.s8_to_f16_array (0 ms)
[ RUN      ] FastNumericConversion.u8_to_f16_array
[       OK ] FastNumericConversion.u8_to_f16_array (0 ms)
[ RUN      ] FastNumericConversion.u8_to_bf16_array
[       OK ] FastNumericConversion.u8_to_bf16_array (0 ms)
[ RUN      ] FastNumericConversion.s8_to_bf16_array
[       OK ] FastNumericConversion.s8_to_bf16_array (0 ms)
[----------] 5 tests from FastNumericConversion (153 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (153 ms total)
[  PASSED  ] 5 tests.
```

# Small Improvements in This PR (Pushing as we Review)
- [x] `FragmentShuffler` needs to specialized for matrix A and matrix B. It currently works when 8b integers are coming from matrix B. 

# Future Improvements (To be taken up in seperate new PRs)
I believe there is more optimal sequence of PTX instructions to achieve the following:
- `S8`->`BF16` conversion can get better and I am working on it. This particular component is a partial specialization `struct FastNumericArrayConverter<cutlass::bfloat16_t, int8_t, 4, Round>` in `numeric_converter.h`. 
- The number of `SFHL` instructions to achieve layout conversion by `FragmentShuffler`could _potentially_ be reduced by increasing a bit of register pressure. 
We can take up these potential improvements in the later PRs.

### CODE REVIEW

File: test/unit/gemm/warp/CMakeLists.txt [hunk 1]
```diff
@@ -28,18 +28,19 @@
...
+  #gemm_sm80.cu
+  #gemm_complex_sm80.cu
+  #gemm_sparse_sm80.cu
+  #gemm_gaussian_complex_sm80.cu
+  gemm_mixed_input_sm80.cu
+  #gemm_sm90.cu
+  #gemm_complex_sm90.cu
+  #wmma_sm70.cu
+  #wmma_sm72.cu
+  #wmma_sm75.cu
```
Comment (manishucsd) [inline test/unit/gemm/warp/CMakeLists.txt:45]: RESOLVE THIS: Uncomment it before the merge. These are commented out for speed of development and reducing compilation times.

File: tools/library/CMakeLists.txt [hunk 2]
```diff
@@ -67,19 +67,20 @@ cutlass_add_library(
...
+  #src/reference/gemm_int8_interleaved_64.cu
+  #src/reference/gemm_e4m3a_e4m3out.cu
+  #src/reference/gemm_e5m2a_e4m3out.cu
+  #src/reference/gemm_e4m3a_e5m2out.cu
+  #src/reference/gemm_e5m2a_e5m2out.cu
+  #src/reference/gemm_fp8in_fp16out.cu
+  #src/reference/gemm_fp8in_bf16out.cu
+  #src/reference/gemm_fp8in_fp32out.cu
+  #src/reference/gemm_fp32out.cu
+  #src/reference/gemm_fp_other.cu
```
Comment (manishucsd) [inline tools/library/CMakeLists.txt:82]: RESOLVE THIS: Uncomment it before the merge. These are commented out for speed of development and reducing compilation times.

File: tools/library/src/reference/initialize_reference_operations.cu [hunk 3]
```diff
@@ -43,45 +43,47 @@ namespace cutlass {
...
+// void initialize_gemm_reference_operations_int8_canonical(Manifest &manifest);
+// void initialize_gemm_reference_operations_e4m3a_e4m3out(Manifest &manifest);
+// void initialize_gemm_reference_operations_e5m2a_e4m3out(Manifest &manifest);
+// void initialize_gemm_reference_operations_e4m3a_e5m2out(Manifest &manifest);
+// void initialize_gemm_reference_operations_e5m2a_e5m2out(Manifest &manifest);
+// void initialize_gemm_reference_operations_fp8in_fp16out(Manifest &manifest);
+// void initialize_gemm_reference_operations_fp8in_bf16out(Manifest &manifest);
+// void initialize_gemm_reference_operations_fp8in_fp32out(Manifest &manifest);
+// void initialize_gemm_reference_operations_fp32out(Manifest &manifest);
+// void initialize_gemm_reference_operations_fp_other(Manifest &manifest);
```
Comment (manishucsd) [inline tools/library/src/reference/initialize_reference_operations.cu:58]: RESOLVE THIS: Uncomment it before the merge. These are commented out for speed of development and reducing compilation times.

File: tools/library/src/reference/initialize_reference_operations.cu [hunk 4]
```diff
@@ -43,45 +43,47 @@ namespace cutlass {
...
+  // initialize_gemm_reference_operations_e4m3a_e4m3out(manifest);
+  // initialize_gemm_reference_operations_e5m2a_e4m3out(manifest);
+  // initialize_gemm_reference_operations_e4m3a_e5m2out(manifest);
+  // initialize_gemm_reference_operations_e5m2a_e5m2out(manifest);
+  // initialize_gemm_reference_operations_fp8in_fp16out(manifest);
+  // initialize_gemm_reference_operations_fp8in_bf16out(manifest);
+  // initialize_gemm_reference_operations_fp8in_fp32out(manifest);
+
+  // initialize_gemm_reference_operations_fp32out(manifest);
+  // initialize_gemm_reference_operations_fp_other(manifest);
```
Comment (manishucsd) [inline tools/library/src/reference/initialize_reference_operations.cu:85]: RESOLVE THIS: Uncomment it before the merge. These are commented out for speed of development and reducing compilation times.

**Comment by rhenry-nv** [comment]
Hi @manishucsd,

Thanks for the awesome work! One high level question - do you plan to handle the scales in the epilogue?

The FT implementation does:
```
half_t converted_val = i2f(val)
half_t mma_input_a = scale * converted_val
acc += mma(mma_input_a, mma_input_b)
```

Even with FP32 accum, we saw that model accuracy can degrade if we scale in the epilogue. I think it is because the input range of `mma_input_a` is much larger if we do mma with it unscaled.

File: include/cutlass/arch/mma.h [hunk 5]
```diff
@@ -68,14 +68,24 @@ struct OpMultiplyAddFastF16 {};
 
 /////////////////////////////////////////////////////////////////////////////////////////////////
 
+/// Tag indicating the input data types are mixed and the narrower type is 
+/// converted to the wider type
+struct OpMultiplyAddMixedInput {};
```
Comment (rhenry-nv) [inline include/cutlass/arch/mma.h:73]: Can we call this `OpMultiplyAddMixedInputUpcast`? It should leave room if we implement a version that converts the wider type to the narrower type

File: include/cutlass/numeric_conversion.h [hunk 6]
```diff
@@ -2441,6 +2442,225 @@ struct FastNumericArrayConverter<int8_t, float, N, Round> {
...
+    return result;
+  }
+
+  CUTLASS_DEVICE
+  result_type operator()(source_type const &s) const { 
+    return convert(s);
+  }
+};
+
+/// Partial specialization for Array<cutlass::bfloat16_t, 4> <= Array<int8_t, 4>
```
Comment (gflegar) [inline include/cutlass/numeric_conversion.h:2586]: What is the complete PTX / SASS that we end up generating here?

What we used when implementing this conversion in Triton is somewhat different, see openai/triton#1879.

Basically, we emit this PTX:

```
    "mov.b32 {s0, s1, s2, s3}, $2;               \n" // unpack
    "cvt.rn.f32.s8 f0, s0;                       \n" // no s8->bf16 pre-Hopper
    "cvt.rn.f32.s8 f1, s1;                       \n" // fi[0:15] is always 0
    "cvt.rn.f32.s8 f2, s2;                       \n" //
    "cvt.rn.f32.s8 f3, s3;                       \n" //
    "prmt.b32 $0, f0, f1, 0x7632;                \n" // f32->bf16 + pack
    "prmt.b32 $1, f2, f3, 0x7632;                \n" //
```

Are you expecting the conversion used here to be more efficient than what is in Triton?

File: include/cutlass/arch/mma.h (hunk 5 above)
Comment (manishucsd) [reply include/cutlass/arch/mma.h]: Thanks, this makes sense. I have updated the name.

File: include/cutlass/numeric_conversion.h (hunk 6 above)
Comment (manishucsd) [reply include/cutlass/numeric_conversion.h:2586]: The PTX for the conversion here is as follows:
```
	prmt.b32 %r347,%r315,%r315,43682;
	prmt.b32 %r349,%r315,%r315,48051;
	add.s32 	%r693, %r343, 1262485504;
	mov.b32 	%f689, %r693;
	add.f32 	%f530, %f689, 0fCB400000;
	add.s32 	%r694, %r345, 1262485504;
	mov.b32 	%f690, %r694;
	add.f32 	%f529, %f690, 0fCB400000;
	add.s32 	%r695, %r347, 1262485504;
	mov.b32 	%f691, %r695;
	add.f32 	%f532, %f691, 0fCB400000;
	add.s32 	%r696, %r349, 1262485504;
	mov.b32 	%f692, %r696;
	add.f32 	%f531, %f692, 0fCB400000;
	cvt.rn.bf16x2.f32 %r351, %f529, %f530;
	cvt.rn.bf16x2.f32 %r352, %f531, %f532;
```

SASS is as follows:
```
                 PRMT R19, R4, 0xbbb3, R4 ;
                 IADD3 R19, R19, 0x4b400000, RZ ;
                 FADD R19, R19, -12582912 ;
                 F2FP.BF16.PACK_AB R57, R19, R10 ;
                 IADD3 R24, R24, 0x4b400000, RZ ;
                 FADD R24, R24, -12582912 ;
                 F2FP.BF16.PACK_AB R11, R7, R24 ;
```

This extracts and converts _two_ `s8`s packed into `s8x4` 32b registers and pack it into `bf16x2` 32b register. 

@gflegar, What is the SASS from the snippet you posted above?

**Comment by manishucsd** [comment]
> Hi @manishucsd,
> 
> Thanks for the awesome work! One high level question - do you plan to handle the scales in the epilogue?
> 
> The FT implementation does:
> 
> ```
> half_t converted_val = i2f(val)
> half_t mma_input_a = scale * converted_val
> acc += mma(mma_input_a, mma_input_b)
> ```
> 
> Even with FP32 accum, we saw that model accuracy can degrade if we scale in the epilogue. I think it is because the input range of `mma_input_a` is much larger if we do mma with it unscaled (since scale is usually much smaller than 1).

Thanks for bringing this to my attention. I am not planning to handle this requirement in the current PR, just mixed-input upcast GEMMs. It should be passing additional scalar argument to the mainloop and let us handle it in another PR?

File: tools/library/scripts/library.py [hunk 7]
```diff
@@ -935,8 +937,13 @@ def CalculateSmemUsage(operation):
                      cta_shape[0] * (cta_shape[2] // 2) // elements_per_8b_md
   else:
     # Few BLAS3 operations only have A tensor
-    smem_per_stage = DataTypeSize[operation.A.element] * cta_shape[0] * cta_shape[2] // 8 + \
-                     DataTypeSize[operation.A.element] * cta_shape[1] * cta_shape[2] // 8
+    data_type_size_a = DataTypeSize[operation.A.element]
```
Comment (hwu36) [inline tools/library/scripts/library.py:940]: this part is for blas3.

File: tools/library/src/reference/gemm_fp_other.cu [hunk 8]
```diff
@@ -77,6 +77,24 @@ void initialize_gemm_reference_operations_fp_other(Manifest &manifest) {
     complex<double>,
     complex<double>
   >(manifest);
+
```
Comment (hwu36) [inline tools/library/src/reference/gemm_fp_other.cu:80]: these should not put in `gemm_fp_mixed_input.cu`?

Comment (manishucsd) [reply tools/library/src/reference/gemm_fp_other.cu]: Yes, I added the mixed input gemm reference in a new file and forgot to remove it from here during the dev. Thanks for catching it. Resolved!

File: include/cutlass/numeric_conversion.h [hunk 9]
```diff
@@ -2441,6 +2442,225 @@ struct FastNumericArrayConverter<int8_t, float, N, Round> {
...
+    // Vectorized s8->f16 conversion using packed instructions
+    uint32_t const* source_ptr = reinterpret_cast<uint32_t const*>(&source);
+    uint32_t* result_ptr = reinterpret_cast<uint32_t*>(&result);
+
+    // Pack s8x2 (s8[1], s8[0]) -> s16x2 (sext.s8[1], sext.s8[0])
+    // (See https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-prmt)
+    // The inline ptx below uses `msb=0` and `msb=1` from the above link to sext the sign-bit in 0, 1, 2, 3 bytes of s8x4
+    // into result_ptr[0] and result_ptr[1]'s 08-15 and 24-31 bits, respectively.
+    // Note that `__byte_perm(source_ptr[0], source_ptr[0], 0x9180);` won't acheive the same and doesn't sext the sign-bit.
+    // Thus, we use inline ptx `prmt.b32` instruction for the desired sext of s8x2 to s16x2.
```
Comment (jackkosaian) [inline include/cutlass/numeric_conversion.h:2475]:
```suggestion
    // Note that `__byte_perm(source_ptr[0], source_ptr[0], 0x9180);` won't acheive the same and doesn't set the sign-bit.
    // Thus, we use inline ptx `prmt.b32` instruction for the desired set of s8x2 to s16x2.
```
Typo? Or is it short for sign extension/extend?

Comment (manishucsd) [reply include/cutlass/numeric_conversion.h:2475]: sext = sign-extend here.

**Review by hwu36 — APPROVED** [review]

**Comment by alexsamardzic** [comment]
> > Hi @manishucsd,
> > Thanks for the awesome work! One high level question - do you plan to handle the scales in the epilogue?
> > The FT implementation does:
> > ```
> > half_t converted_val = i2f(val)
> > half_t mma_input_a = scale * converted_val
> > acc += mma(mma_input_a, mma_input_b)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Even with FP32 accum, we saw that model accuracy can degrade if we scale in the epilogue. I think it is because the input range of `mma_input_a` is much larger if we do mma with it unscaled (since scale is usually much smaller than 1).
> 
> Thanks for bringing this to my attention. I am not planning to handle this requirement in the current PR, just mixed-input upcast GEMMs. It should be passing additional scalar argument to the mainloop and let us handle it in another PR?

Thanks for adding mixed GEMM support!  I have the same question as OP: Any follow-ups on implementing scales, and also how about (u)int4 support (packed by two into an 8-bit integer)?

**Pull request merged**